src/common/debugAdapter.ts

```diff
@@ -6,147 +6,856 @@
 
 import * as vscode from 'vscode';
 import { DebugProtocol } from '@vscode/debugprotocol';
-
+import { ServicePseudoTerminal, TerminalMode } from '@vscode/sync-api-service';
+import { Spawnee, Spawner } from './spawner';
 import RAL from './ral';
-import { Event, Response } from './debugMessages';
-import { Launcher } from './launcher';
-import { ServicePseudoTerminal } from '@vscode/sync-api-service';
-import { Terminals } from './terminals';
-
-export class DebugAdapter implements vscode.DebugAdapter {
 
-	private readonly context: vscode.ExtensionContext;
+const StackFrameRegex = /^[>,\s]+(.+)\((\d+)\)(.*)\(\)/;
+const ScrapeDirOutputRegex = /\[(.*)\]/;
+const BreakpointRegex = /Breakpoint (\d+) at (.+):(\d+)/;
+const PossibleStepExceptionRegex = /^\w+:\s+.*\r*\n>/;
+const PrintExceptionMessage = `debug_pdb_print_exc_message`;
+const SetupExceptionMessage = `alias debug_pdb_print_exc_message !import sys; print(sys.exc_info()[1])`;
+const PrintExceptionTraceback = `debug_pdb_print_exc_traceback`;
+const SetupExceptionTraceback = `alias debug_pdb_print_exc_traceback !import traceback; import sys; traceback.print_exception(*sys.exc_info())`;
+const PdbTerminator = `(Pdb) `;
 
-	private sequenceNumber: number;
-	private _sendMessage: vscode.EventEmitter<vscode.DebugProtocolMessage>;
+export class DebugAdapter implements vscode.DebugAdapter {
+	private _launcher: () => void;
+	private _cwd: string | undefined;
+	private _sequence = 0;
+	private _spawner: Spawner;
+	private _debuggee: Spawnee | undefined;
+	private _disposables: vscode.Disposable[] = [];
+	private _outputChain: Promise<string> | undefined;
+	private _outputEmitter = new vscode.EventEmitter<string>();
+	private _stopped = true;
+	private _stopOnEntry = false;
+	private _currentFrame = 1;
+	private _disposed = false;
+	private _uncaughtException = false;
+	private _workspaceFolder: vscode.WorkspaceFolder | undefined;
+	private _boundBreakpoints: DebugProtocol.Breakpoint[] = [];
+	private _didSendMessageEmitter: vscode.EventEmitter<DebugProtocol.Response | DebugProtocol.Event> =
+		new vscode.EventEmitter<DebugProtocol.Response | DebugProtocol.Event>();
 
-	private launcher: Launcher | undefined;
+	private static _terminal: vscode.Terminal;
+	private static _debugTerminal: ServicePseudoTerminal<any>;
 
-	constructor(_vscodeSession: vscode.DebugSession, context: vscode.ExtensionContext) {
-		this.context = context;
-		this.sequenceNumber = 1;
-		this._sendMessage = new vscode.EventEmitter();
-		this.onDidSendMessage = this._sendMessage.event;
+	constructor(
+		readonly session: vscode.DebugSession,
+		readonly context: vscode.ExtensionContext,
+		private readonly _ral: RAL
+	) {
+		this._spawner = _ral.spawner.create();
```

Comment:
Debugger is using a 'Spawner' and 'Spawnee' abstraction. It needs only a few things from child_process.

Reply:
Once the WASI version is working, I hope this should be easy to implement there.